net: ipv4: Make native IPv4 options dependent on NET_NATIVE_IPV4

rlubos opened the pull request:
IPv4 Kconfig options which only affect native IPv4 stack should be dependent on `NET_NATIVE_IPV4`, similarly as it's done for IPv6.

rlubos committed:
net: ipv4: Make native IPv4 options dependent on NET_NATIVE_IPV4

IPv4 Kconfig options which only affect native IPv4 stack should be
dependent on NET_NATIVE_IPV4, similarly as it's done for IPv6.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
@jukkar Currently this patch follows what's been done for IPv6:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/Kconfig.ipv6

I've noticed one inconsequence with this approach though, hence marking this as a draft for now. 

According to Kconfig, the IPv4/6 address counts does not depend on native IP networking. Following this logic, it should be possible to register IP addresses even for offloaded interfaces, which IMO makes sense, as for example we should be able then to display them with net shell.

However, when I look how those IPv4/6 address containers are defined, they only seem to be enabled if native networking is used:
https://github.com/zephyrproject-rtos/zephyr/blob/119b238594971a41dc55609d7509677efc675a6c/subsys/net/ip/net_if.c#L98

Shouldn't those rather depend on `CONFIG_NET_IPV4/6` instead, so that it's still possible to register IP addresses in purely offloaded builds? I'm pretty sure registering IP addresses for offloaded interfaces works, as we do that for nRF91 in some NCS samples. It requires though native networking support enabled, which IMO shouldn't be the case.

I can investigate this further if we agree this should be changed.

jukkar commented:
Good point, I agree, we can have the addresses for offloaded interfaces

rlubos committed:
net: iface: Don't require native networking to register IP address

It shouldn't be needed to enable native networking to be able to
add/delete IP addresses on network interfaces.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
test: net: offloaded_netdev: Add tests for IPv4/6 address registration

Make sure it's possible to register IPv4/6 addresses on an offloaded
interface. Add an extra build configuration to make sure it's also
possible when native IP stack is disabled.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
Ok, I have updated `net_if.c` ifdefs so that it's now possible to register IP addresses even if native IP stack is disabled. This allows for example to use connection manager features w/o a need to enable native stack in purely offloaded scenarios. I think it's ready to review.

I've moved a few functions in the file (no functional changes just relocating the code) to avoid further ifdef-slicling of the file. Alternatively, I can keep the code as is, and introduce more ifdeffing in case that's preferred.

rlubos commented:
Latest push with CI fixes.

jukkar commented:
There are also checks for native IPv4/6 in `ipv4.c` and `ipv6.c` in net-shell, perhaps those should be adjusted too.

rlubos committed:
net: shell: Enable IPv4/6 and iface commands if NET_NATIVE is disabled

Some commands can be executed and some statuses can be printed even if
native IP is disabled.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

pdgendt commented on subsys/net/lib/shell/iface.c:
```diff
@@ -148,17 +147,17 @@ static const char *iface_flags2str(struct net_if *iface)
-#if defined(CONFIG_NET_IPV6)
-	struct net_if_ipv6_prefix *prefix;
+#if defined(CONFIG_NET_NATIVE_IPV6)
+	struct net_if_ ipv6_prefix *prefix;
```
Looks like a typo

rlubos replied:
yep, thanks!

rlubos commented:
> There are also checks for native IPv4/6 in `ipv4.c` and `ipv6.c` in net-shell, perhaps those should be adjusted too.

Good point @jukkar, I've modified `net iface/ipv4/ipv6` commands so that they can work and print statuses with native IP stack disabled. Although, to make net shell functional w/o native stack, we need your shell PR merged first, so that the disabled shell modules are not compiled in, currently some of them throw errors if built w/o native stack.
